db/validation.go

```diff
@@ -0,0 +1,85 @@
+package db
+
+import (
+	"bytes"
+	"net/http"
+	"strings"
+
+	"github.com/couchbase/sync_gateway/base"
+)
+
+// validateNewBody validates any new body being received (i.e. through blip, import, and API)
+func validateNewBody(body Body) error {
+	// Reject a body that contains the "_removed" property, this means that the user
+	// is trying to update a document they do not have read access to.
+	if body[BodyRemoved] != nil {
+		return base.HTTPErrorf(http.StatusNotFound, "Document revision is not accessible")
+	}
+
+	// Reject bodies that contains the "_purged" property.
+	if body[BodyPurged] != nil {
+		return base.HTTPErrorf(http.StatusBadRequest, "user defined top-level property '_purged' is not allowed in document body")
+	}
+
+	for key := range body {
+		if strings.HasPrefix(key, BodyInternalPrefix) {
```

Comment:
I think this is relevant even for existing code doing this byte/string checking upfront - but these checks are case-sensitive, and the Go JSON implementation is case _insensitive._

That difference might be the cause of subtle bugs, if a user is able to bypass the validation by writing properties in uppercase.

Reply:
Will do as part of [CBG-2051](https://issues.couchbase.com/browse/CBG-2051)